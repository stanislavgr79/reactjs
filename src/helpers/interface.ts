export interface IMovie {
  [x: string]: any;
  id?: number;
  title: string;
  tagline?: string;
  vote_average?: number;
  vote_count?: number;
  release_date: string;
  movieUrl: string;
  genres: string[] | { label: string, value: string }[];
  poster_path?: string | undefined;
  overview: string;
  runtime: number | string | undefined;
}
