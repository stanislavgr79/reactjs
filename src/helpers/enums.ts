/* eslint-disable no-unused-vars */

/**
 * Enum for common SelectGenres.
 * @readonly
 * @enum {{label: string, value: string}}
 */
export const selectGenres = Object.freeze([
  { label: 'Action', value: 'Action' },
  { label: 'Adventure', value: 'Adventure' },
  { label: 'Animation', value: 'Animation' },
  { label: 'Comedy', value: 'Comedy' },
  { label: 'Documentary', value: 'Documentary' },
  { label: 'Drama', value: 'Drama' },
  { label: 'Family', value: 'Family' },
  { label: 'Horror', value: 'Horror' },
  { label: 'Mystery', value: 'Mystery' },
  { label: 'Thriller', value: 'Thriller' },
]);

/**
 * Enum for common Genres.
 * @readonly
 * @enum { string }
 */
export enum Genres {
  ALL = '',
  DOCUMENTARY = 'documentary',
  COMEDY = 'comedy',
  HORROR = 'horror',
  CRIME = 'crime',
}

/**
 * Enum for common SortBy.
 * @readonly
 * @enum { string }
 */
export enum SortBy {
  TITLE = 'title',
  RELEASE = 'release_date',
  RUNTIME = 'runtime',
  RATING = 'vote_average',
}

/**
 * Enum for common SortOrder.
 * @readonly
 * @enum { string }
 */
export enum SortOrder {
  ASC = 'asc',
  DESC = 'desc',
}
