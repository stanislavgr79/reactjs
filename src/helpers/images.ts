const images = [
  {
    name: 'Ricky and Morty.jpg',
    value: './resources/images/Ricky and Morty.jpg',
  },
  {
    name: 'Pulp Fiction.jpg',
    value: './resources/images/Pulp Fiction.jpg',
  },
  {
    name: 'Kill Bill.jpg',
    value: './resources/images/Kill Bill.jpg',
  },
  {
    name: 'Avengers.jpg',
    value: './resources/images/Avengers.jpg',
  },
  {
    name: 'Inception.jpg',
    value: './resources/images/Inception.jpg',
  },
  {
    name: 'Reservoir Dogs.jpg',
    value: './resources/images/Reservoir Dogs.jpg',
  },
];

export default images;
