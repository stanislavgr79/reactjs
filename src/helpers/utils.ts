export const createClassList = (classesArray: string[]) =>
  classesArray
    .filter((className) => typeof className === 'string' && className.length > 0)
    .join(' ')
    .trim();

export function transformGenresToStringArray(genres: { value: string, label: string }[]) {
  const genresArray: string[] = [];
  genres.map((it) => genresArray.push(it.value));
  return genresArray;
}

export function transformGenresToArrayObjects(genres: string[]) {
  const genresArray: { value: string, label: string }[] = [];
  genres.map((it: string) =>
    genresArray.push({
      label: it,
      value: it,
    }),
  );
  return genresArray;
}
