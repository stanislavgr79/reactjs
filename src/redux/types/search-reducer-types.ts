export const UPDATE_SEARCH_QUERY = 'UPDATE_SEARCH_QUERY';
export const UPDATE_LOCATION = 'UPDATE_LOCATION';

export interface SearchState {
  search: {
    value: string,
  };
  location: string;
}

interface UpdateSearchQuery {
  type: typeof UPDATE_SEARCH_QUERY;
  search: {
    value: string,
  };
}

interface UpdateLocation {
  type: typeof UPDATE_LOCATION;
  payload: string;
}

export type SearchActionTypes = UpdateSearchQuery | UpdateLocation;
