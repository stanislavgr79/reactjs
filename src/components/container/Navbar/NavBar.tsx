import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../../redux/redux-store';

import SidebarListNav from '../../../components/atom/SideBarListNav';
import SidebarSelectNav from '../../../components/atom/SidebarSelectNav';
import Button from '../../atom/Button';

import { Genres, SortBy, SortOrder } from '../../../helpers/enums';
import {
  updateCurrentGenre,
  updateCurrentSortBy,
  updateSortOrder,
} from '../../../redux/actions/sidebar-actions';

import './NavBar.scoped.scss';

export default function NavBar(): JSX.Element {
  const dispatch = useDispatch();
  const { genre, sortBy, sortOrder } = useSelector((store: AppState) => {
    return store.sidebar;
  });
  const [sortOrderState, setSortOrderState] = useState(sortOrder);

  useEffect(() => {
    dispatch(updateSortOrder(sortOrderState));
  }, [dispatch, sortOrderState]);

  const handleChangeGenre = useCallback(
    (genre: Genres) => {
      dispatch(updateCurrentGenre(genre));
    },
    [dispatch],
  );

  const handleChangeSortBy = useCallback(
    (sortBy: SortBy) => {
      dispatch(updateCurrentSortBy(sortBy));
    },
    [dispatch],
  );

  const changeSortOrder = () => {
    sortOrderState == SortOrder.ASC
      ? setSortOrderState(SortOrder.DESC)
      : setSortOrderState(SortOrder.ASC);
  };

  return (
    <div className="nav-bar container-md">
      <div className="side-genre">
        <SidebarListNav
          defaultValue={genre}
          updateCurrentGenre={(genre: Genres) => handleChangeGenre(genre)}
        />
      </div>
      <div className="side-sort ">
        <label>SORT BY</label>
        <SidebarSelectNav
          defaultValue={sortBy}
          updateCurrentSortBy={(sortBy: SortBy) => handleChangeSortBy(sortBy)}
        />
        <div className="sort_order">
          <Button
            buttonType="button"
            className={`btn_arrow ${sortOrderState == SortOrder.ASC ? `arrow_down` : `arrow_up`}`}
            onClick={changeSortOrder}
            title="Change sort order"
          />
        </div>
      </div>
    </div>
  );
}
