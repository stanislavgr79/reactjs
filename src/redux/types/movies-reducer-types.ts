import { IMovie } from '../../helpers/interface';

export const ADD_TO_MOVIES = 'ADD_TO_MOVIES';
export const DELETE_FROM_MOVIES = 'DELETE_FROM_MOVIES';
export const UPDATE_MOVIE_IN_MOVIES = 'UPDATE_MOVIE_IN_MOVIES';
export const UPDATE_STATUS_MOVIE_POPUP = 'UPDATE_STATUS_MOVIE_POPUP';
export const UPDATE_SELECTED_MOVIE_ID = 'UPDATE_SELECTED_MOVIE_ID';

export interface MoviesState {
  movies: IMovie[];
  moviesIsLoading: boolean;
  showPopup: boolean;
  selectedMovieId: number;
  movie: IMovie | undefined;
}

interface AddMovieToMovies {
  type: typeof ADD_TO_MOVIES;
  payload: IMovie;
}

interface DeleteFromMovies {
  type: typeof DELETE_FROM_MOVIES;
  payload: number;
}

interface UpdateMovieInMovies {
  type: typeof UPDATE_MOVIE_IN_MOVIES;
  payload: IMovie;
}

interface UpdateStatusMoviePopup {
  type: typeof UPDATE_STATUS_MOVIE_POPUP;
  payload: boolean;
}

interface UpdateSelectedMovieId {
  type: typeof UPDATE_SELECTED_MOVIE_ID;
  payload: number;
}

export type MoviesActionTypes =
  | AddMovieToMovies
  | DeleteFromMovies
  | UpdateMovieInMovies
  | UpdateStatusMoviePopup
  | UpdateSelectedMovieId;
