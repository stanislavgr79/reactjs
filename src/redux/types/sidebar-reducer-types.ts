import { Genres, SortBy, SortOrder } from '../../helpers/enums';

/* eslint-disable no-unused-vars */
export const UPDATE_SELECT_GENRE = 'UPDATE_SELECT_GENRE';
export const UPDATE_SORT_BY = 'UPDATE_SORT_BY';
export const UPDATE_SORT_ORDER = 'UPDATE_SORT_ORDER';
export const UPDATE_SELECTED_INDEX_LIST_NAV = 'UPDATE_SELECTED_INDEX_LIST_NAV';

export interface SidebarState {
  genre: Genres;
  sortBy: SortBy;
  sortOrder: SortOrder;
  selectedIndexListNav?: number;
}

interface UpdateCurrentGenre {
  type: typeof UPDATE_SELECT_GENRE;
  payload: Genres;
}

interface UpdateCurrentSortBy {
  type: typeof UPDATE_SORT_BY;
  payload: SortBy;
}

interface UpdateSortOrder {
  type: typeof UPDATE_SORT_ORDER;
  payload: SortOrder;
}

interface UpdateSelectedIndexListNav {
  type: typeof UPDATE_SELECTED_INDEX_LIST_NAV;
  payload: number;
}

export type SidebarActionTypes =
  | UpdateCurrentGenre
  | UpdateCurrentSortBy
  | UpdateSortOrder
  | UpdateSelectedIndexListNav;
