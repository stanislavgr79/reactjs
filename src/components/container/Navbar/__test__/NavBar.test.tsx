import React from 'react';
import { Provider } from 'react-redux';

import NavBar from '../NavBar';

import configureStore from 'redux-mock-store';
import { render, screen, fireEvent } from '@testing-library/react';
import storeReal from '../../../../redux/redux-store';
import { SidebarState } from '../../../../redux/types/sidebar-reducer-types';
import { Genres, SortBy, SortOrder } from '../../../../helpers/enums';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';

jest.mock('../../../atom/SideBarListNav', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="sidebar-list-nav"></div>;
    },
  };
});
jest.mock('../../../atom/SidebarSelectNav', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="sidebar-select-nav"></div>;
    },
  };
});
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

interface AppState {
  sidebar: SidebarState;
}
const initialState: AppState = {
  sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
};
const buildStore = configureStore([thunk]);
let store = buildStore(initialState);

jest.mock('../../../../redux/actions/sidebar-actions', () => ({
  updateSortOrder: () => {
    console.log('!!!!!!!!!!!!!!!!!!!!!!!!');
    const newInitialState: AppState = {
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.DESC },
    };
    store = buildStore(newInitialState);
    console.log('!!!!!!!!!!!!!!!!!!!!!!!!');
  },
}));

describe('NavBar', () => {
  // let store;
  // let wrapper;

  beforeEach(() => {
    store = buildStore(initialState);
    // wrapper = shallow(<colorbuttons store="{store}" />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // test('snapshot renders', () => {
  //   const tree = render(
  //     <Provider store={store}>
  //       <NavBar />
  //     </Provider>,
  //   );

  //   expect(tree.baseElement).toMatchSnapshot();
  //   expect(tree.baseElement).toContainHTML('<div class="nav-bar container-md">');
  // });

  test('snapshot renders2', () => {
    // const store = storeReal;
    // store = buildStore(initialState);
    // const spy = jest.spyOn(store, 'dispatch');

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <NavBar />
        </Provider>
      </BrowserRouter>,
    );
    // const wrapper = shallow(
    //   <Provider store={store}>
    //     <NavBar />
    //   </Provider>,
    // );

    fireEvent.click(screen.getByTitle('Change sort order'));
    expect(tree.baseElement).toMatchSnapshot();
    expect(tree.baseElement).toContainHTML('<div class="nav-bar container-md">');
    // expect(spy).toHaveBeenCalled();
    console.log(store.getState());
  });
});
