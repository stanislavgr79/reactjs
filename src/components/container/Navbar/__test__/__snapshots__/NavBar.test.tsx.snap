// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar snapshot renders2 1`] = `
<body>
  <div>
    <div
      class="nav-bar container-md"
    >
      <div
        class="side-genre"
      >
        <div
          class="sidebar-list-nav"
        />
      </div>
      <div
        class="side-sort "
      >
        <label>
          SORT BY
        </label>
        <div
          class="sidebar-select-nav"
        />
        <div
          class="sort_order"
        >
          <button
            class="btn button btn_arrow arrow_up"
            id=""
            title="Change sort order"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
